resources/google_agent.rb

```diff
@@ -0,0 +1,177 @@
+    bash node['file_url_name'] do
+      cwd node['tmp_file_path']
+      code "./#{node['file_url_name']} --also-install --version=#{node['version']}"
+      only_if "./#{node['file_url_name']} --also-install --version=#{node['version']} --dry-run | grep -qi 'installation succeeded'" == 0
+      user 'root'
+    end
+    case node['final_agent_type']
+    when 'google-cloud-ops'
+      servicename = 'google-cloud-ops-agent'
+    when 'monitoring'
+      servicename = 'stackdriver-agent'
+    when 'logging'
+      servicename = 'google-fluentd'
+    end
+    service servicename do
+      action [:start, :enable]
+      # TODO: When CINC fixes the bug related to this, uncomment the below line
+      # user 'root'
+    end
+    # The user parameter for service isn't working, hack to bypass
+    # TODO: once the service command above has the bug fixed and can
+    # specify user properly, remove this.
+    bash 'hack service' do
+      user 'root'
+      code "systemctl enable #{servicename}"
+    end
+  when 'windows'
+    powershell_script node['file_url_name'] do
+      cwd node['tmp_file_path']
+      code "#{node['tmp_file_path']}/#{node['file_url_name']} -AlsoInstall -Version #{node['version']}"
+      only_if "#{node['file_url_name']} -AlsoInstall -Version #{node['version']} -WhatIf | findstr /i \"Verification of google-cloud-ops-agent.x86_64.${version} completed\"" == 0
+    end
+  end
+end
+
+# Uninstall the software, and remove the repository
+action :uninstall do
+  # Run the appropriate uninstaller
+  case node['platform_family']
+  when 'rhel', 'debian', 'suse', 'amazon'
+    bash node['file_url_name'] do
+      cwd node['tmp_file_path']
+      code "./#{node['file_url_name']} --uninstall --remove-repo"
+      only_if "./#{node['file_url_name']} --uninstall --remove-repo --dry-run | grep -qi 'uninstallation succeeded'" == 0
+      unless node['file_url_name'].include? 'ops-agent'
+        action :nothing
+      end
+      user 'root'
+      notifies :run, 'bash[daemon-reload]', :immediately
+    end
+    # Script does not stop logging or monitoring services
+    # So we must do it manually
+    case node['final_agent_type']
+    when 'google-cloud-ops'
+      servicename = 'google-cloud-ops-agent'
+    when 'monitoring'
+      servicename = 'stackdriver-agent'
+    when 'logging'
+      servicename = 'google-fluentd'
+    end
+    service servicename do
+      action :disable
+      # TODO: When CINC fixes the bug related to this, uncomment the below line
+      # user 'root'
+    end
+    service servicename do
+      action :stop
+      notifies :run, "bash[#{node['file_url_name']}]", :immediately
+      # TODO: When CINC fixes the bug related to this, uncomment the below line
+      # user 'root'
+    end
+    bash 'daemon-reload' do
+      code 'systemctl daemon-reload && systemctl reset-failed'
+      user 'root'
+      action :nothing
+    end
+    # Uninstall is not properly removing these files, so we do it here
```

Comment:
Does line 142 - 147 need to be conditional based on the values of `package_state` and `agent_type`? Does Ops Agent have similar issue?

Reply:
This is already conditional due to being in the uninstall block. Additionally, rather than making it dependent on agent_type, it only performs the action if the file exists. Should agent_type protection be added in the event they have multiple agent types on a single system?

Reply:
If user has a Chef cookbook that is configured to ensure the logging agent is not present. And the VM already has monitoring agent on it. 

This might be a corner case, but imagine an user has an existing VM that already has the Monitoring Agent running, which is not managed by Chef. If the user specifies some Chef cookbook to manage logging agent and ops agent, this cookbook should not impact service related files that belong to the Monitoring Agent, right? That should only happen if the user explicitly sets `agent_type: monitoring`.

Reply:
I've addressed this in my most recent code. It may be a corner case, but it should be handled correctly.